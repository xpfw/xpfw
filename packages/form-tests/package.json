{
  "name": "@xpfw/form-tests",
  "version": "1.0.0-beta.12",
  "main": "dist",
  "typings": "dist",
  "license": "MIT",
  "peerDependencies": {
    "lodash-es": "^4.x.x",
    "moment": "2.x.x",
    "react": "^16.x.x",
    "@xpfw/form": "1.x.x"
  },
  "dependencies": {
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@types/jest": "^24.0.4",
    "@types/lodash": "^4.14.121",
    "@types/lodash-es": "^4.17.2",
    "@types/mockdate": "^2.0.0",
    "@types/react": "^16.8.3",
    "@types/react-test-renderer": "^16.8.1",
    "@types/socket.io-client": "^1.4.32",
    "lodash-es": "^4.17.11",
    "moment": "^2.24.0",
    "@xpfw/form": "^1.0.0-beta.12",
    "jest": "^24.1.0",
    "jest-cli": "^24.1.0",
    "mockdate": "^2.0.2",
    "react": "^16.8.1",
    "react-test-renderer": "^16.8.2",
    "rollup": "^1.1.2",
    "rollup-plugin-typescript2": "^0.19.2",
    "rollup-plugin-uglify": "^6.0.2",
    "ts-jest": "=23.10.5",
    "tslib": "^1.9.3",
    "typescript": "^3.3.3",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": false
      }
    },
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleNameMapper": {
      "preact-render-to-json": "preact-render-to-json",
      "preact": "preact",
      "react": "preact-compat",
      "react-dom": "preact-compat"
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
