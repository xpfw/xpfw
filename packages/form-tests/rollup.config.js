// rollup.config.js
import typescript from "rollup-plugin-typescript2"
import {uglify} from "rollup-plugin-uglify"
import { minify } from "uglify-es"

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'cjs'
  },
  plugins: [
    typescript(),
    // uglify({}, minify)
  ],
  external: [
    "lodash",
    "moment",
    "mockdate",
    "feathers-client",
    "resub",
    "resub-persist",
    "socket.io-client",
    "@xpfw/validate",
    "@xpfw/form-shared",
    "react",
    "react-dom",
    "preact-render-to-json"
  ]
};