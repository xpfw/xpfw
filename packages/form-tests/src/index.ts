import arrayTest from "./tests/array"
import booleanTest from "./tests/boolean"
import dateTest from "./tests/date"
import locationTest from "./tests/location"
import objectTest from "./tests/object"
import selectTest from "./tests/select"
import stringTest from "./tests/string"
import {
  ArrayField, BooleanField, DateField, LocationField, NameField,
  NumberField, ObjectField, PwField, SelectField
} from "./testUtil/schema"

export {
  booleanTest, stringTest, arrayTest, dateTest,
  locationTest, objectTest, selectTest,
  ArrayField, BooleanField, DateField, LocationField, NameField,
  NumberField, ObjectField, PwField, SelectField
}
