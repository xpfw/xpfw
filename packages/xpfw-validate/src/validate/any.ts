import { get, intersection, isFunction, isNil, isNumber, isObject, isString, keys, set } from "lodash"
import { ErrorType, FieldType, Method } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateArray from "./array"
import validateBool from "./bool"
import validateDate from "./date"
import validateLocation from "./location"
import validateNumber from "./number"
import validateObject from "./object"
import validateQueryObject, { possibleQueryOperators } from "./queryObject"
import validateSelect from "./select"
import validateText from "./text"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateAny: ValidateFunction = (value, field, params) => {
  let errorResolve: any = null
  if (get(params, "method") === Method.Find && isObject(value)) {
    if (intersection(keys(value), possibleQueryOperators).length > 0) {
      const form: any = params.form
      return validateQueryObject(value, form, params)
    }
  }
  if (isObject(field) && !isNil(field.validate)) {
    if (isNumber(field.type)) {
      switch (field.type) {
        case FieldType.Text:
        case FieldType.Password:
        case FieldType.RelationshipSingle: {
          errorResolve = validateText(value, field, params)
          break
        }
        case FieldType.Number:
        case FieldType.Slider: {
          errorResolve = validateNumber(value, field, params)
          break
        }
        case FieldType.Boolean: {
          errorResolve = validateBool(value, field, params)
          break
        }
        case FieldType.Date: {
          errorResolve = validateDate(value, field, params)
          break
        }
        case FieldType.Select: {
          errorResolve = validateSelect(value, field, params)
          break
        }
        case FieldType.Object: {
          errorResolve = validateObject(value, field, params)
          break
        }
        case FieldType.Array:
        case FieldType.RelationshipMulti: {
          if (field.type === FieldType.RelationshipMulti) {
            set(field, "validate.type", FieldType.Text)
          }
          errorResolve = validateArray(value, field, params)
          break
        }
        case FieldType.Location: {
          errorResolve = validateLocation(value, field, params)
          break
        }
        default: {
          errorResolve = null
          break
        }
      }
    } else if (isString(field.type)) {
      if (isObject(params) && isObject(params.registry)) {
        const validator = params.registry.findValidator(field.type)
        if (isFunction(validator)) {
          errorResolve = validator(value, field, params)
        }
      }
    }
    if (isNil(errorResolve)) {
      console.log("erroirrseolve is null for", field)
      errorResolve = Promise.reject(false)
    }
  } else {
    errorResolve = Promise.resolve(value)
  }
  if (isObject(field) && isObject(field.validate)) {
    const validate: any = field.validate
    if (isFunction(validate.custom)) {
      let catched = false
      const errs: any = {mapTo: field.mapTo, errors: []}
      return errorResolve.catch((e: any) => {
        catched = true
        errs.errors = e
        return validate.custom(value, field, params)
      }).then(() => {
        if (!catched) {
          return validate.custom(value, field, params)
        }
        return Promise.reject(errs)
      })
    }
  }
  return errorResolve
}

export default validateAny
