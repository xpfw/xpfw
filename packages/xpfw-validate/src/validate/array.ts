import { cloneDeep, each, get, isArray, isNil, isNumber, isString } from "lodash"
import { ErrorType, FieldType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateAny from "./any"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateArray: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const validArray = isArray(value)
  if (checkShouldBeRequired(field, params) && !validArray) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (validArray) {
    const min = get(field, "validate.min")
    if (isNumber(min)) {
      if (value.length < min) {
        errors.push({
          type: ErrorType.tooLow
        })
      }
    }
    const max = get(field, "validate.max")
    if (isNumber(max)) {
      if (value.length > max) {
        errors.push({
          type: ErrorType.tooHigh
        })
      }
    }
  }
  if (errors.length > 0) {
    return Promise.reject(errors)
  }
  const promises: Array<Promise<any>> = []
  if (!isNil(field.validate)) {
    const subValidate: any = field.validate
    if (isString(subValidate.type) || isNumber(subValidate.type)) {
      each(value, (date) => {
        const tmpFieldOptions = cloneDeep(field)
        tmpFieldOptions.type = subValidate.type
        if (tmpFieldOptions.type === FieldType.Array || tmpFieldOptions.type === FieldType.Object) {
          tmpFieldOptions.validate = subValidate.validate
        }
        promises.push(validateAny(date, tmpFieldOptions, params))
      })
    }
  }
  return Promise.all(promises).then((results) => {
    return value
  })
}

export default validateArray
