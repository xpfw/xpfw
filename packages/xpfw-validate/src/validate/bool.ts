import { isBoolean, isNil } from "lodash"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateBool: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const isValidBoolean = isBoolean(value)
  if (checkShouldBeRequired(field, params) && (!isValidBoolean || !value)) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (!isNil(value) && !isValidBoolean) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateBool
