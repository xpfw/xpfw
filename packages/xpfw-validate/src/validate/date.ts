import { isDate, isNil, isString } from "lodash"
import * as moment from "moment"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"
const m: any = moment
const validateDate: ValidateFunction = (value, field, params) => {
  if (isString(value) && m(value).isValid()) {
    value = m(value).toDate()
  }
  const errors: IErrorBase[] = []
  const isValidDate = isDate(value)
  if (checkShouldBeRequired(field, params) && !isValidDate) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (!isNil(value) && !isValidDate) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateDate
