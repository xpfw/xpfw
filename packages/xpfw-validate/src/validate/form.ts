import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
