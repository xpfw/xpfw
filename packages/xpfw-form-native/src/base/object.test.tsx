import { ComponentRegistry } from "@xpfw/form-shared"
import { tests } from "@xpfw/form-tests"
import { FieldType } from "@xpfw/validate"
import ObjectField from "./object"
import TextField from "./text"

ComponentRegistry.registerComponent(FieldType.Text, TextField)
ComponentRegistry.registerComponent(FieldType.Number, TextField)
ComponentRegistry.registerComponent(FieldType.Object, ObjectField)

test("Object Field Test", tests.object)
