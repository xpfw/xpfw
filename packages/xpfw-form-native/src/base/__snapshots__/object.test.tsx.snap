// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object Field Test: locationField 1`] = `
<View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.myVal"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.myVal",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View />
  </View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.secondNum"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.secondNum",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View />
  </View>
</View>
`;

exports[`Object Field Test: set to invalid 1`] = `
<View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.myVal"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.myVal",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value={NaN}
        />
      </View>
    </View>
    <View />
  </View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.secondNum"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.secondNum",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View />
  </View>
</View>
`;

exports[`Object Field Test: set to otherValidLocation 1`] = `
<View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.myVal"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.myVal",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value={767}
        />
      </View>
    </View>
    <View />
  </View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.secondNum"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.secondNum",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View />
  </View>
</View>
`;

exports[`Object Field Test: set to validLocation 1`] = `
<View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.myVal"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.myVal",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value={41}
        />
      </View>
    </View>
    <View />
  </View>
  <View>
    <Text />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          keyboardType="numeric"
          mapTo="objectKey.secondNum"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          schema={
            Object {
              "title": "objectKey.secondNum",
              "type": "number",
            }
          }
          secureTextEntry={false}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View />
  </View>
</View>
`;
